- reference :// https://pabloinsente.github.io/intro-numpy-fundamentals

In [ ]:
import numpy as np

# 9. Array Indexing
- Numpy array 를 Indexing 하는 다양한 방법을 알아보자.

1) 1차원 array Indexing  
2) 2차원 array Indexing  
3) Indexing and Assignment  
4) Boolean(Masking) Indexing

## 1. 1차원 array Indexing

In [ ]:
array_one = np.arange(1,11)

print(f'array_one: {array_one}')
print(f'array_one.ndim:{array_one.ndim},array_one.shape:{array_one.shape}')

array_one: [ 1  2  3  4  5  6  7  8  9 10]
array_one.ndim:1,array_one.shape:(10,)


In [ ]:
print(f'array_one[0]: {array_one[0]}')
print(f'array_one[5]: {array_one[5]}')
print(f'array_one[9]: {array_one[9]}')
print(f'array_one[-5]: {array_one[-5]}')
print(f'array_one[-1]: {array_one[-1]}')

array_one[0]: 1
array_one[5]: 6
array_one[9]: 10
array_one[-5]: 6
array_one[-1]: 10


**Slicing 1D-array**
- [included:not-included)

In [ ]:
print(f'array_one[0:3]: {array_one[0:3]}')
print(f'array_one[5:9]: {array_one[5:9]}')
print(f'array_one[-9:-5]: {array_one[-9:-5]}')
print(f'array_one[-3:-1]: {array_one[-3:-1]}')
print(f'array_one[3:-1]: {array_one[3:-1]}')

array_one[0:3]: [1 2 3]
array_one[5:9]: [6 7 8 9]
array_one[-9:-5]: [2 3 4 5]
array_one[-3:-1]: [8 9]
array_one[3:-1]: [4 5 6 7 8 9]


**Slicing 1D-array**
- with stride

In [ ]:
print(f'array_one[0:6:1]: {array_one[0:6:1]}')
print(f'array_one[0:6:2]: {array_one[0:6:2]}')
print(f'array_one[-6:-1:3]: {array_one[-6:-1:3]}')

array_one[0:6:1]: [1 2 3 4 5 6]
array_one[0:6:2]: [1 3 5]
array_one[-6:-1:3]: [5 8]


## 2. 2차원 array Indexing
- 1) [,] 방식 -> view
- 2) [][] 방식 -> copy

In [ ]:
array_two = np.arange(1,10).reshape((3,3))
array_three = np.arange(1,9).reshape((2,2,2))

print(f'array_two: \n{array_two}\n')
print(f'array_two.ndim:{array_two.ndim},array_two.shape:{array_two.shape}')
print("\n\n")
print(f'array_three: \n{array_three}\n')
print(f'array_three.ndim:{array_three.ndim},array_three.shape:{array_three.shape}')

array_two: 
[[1 2 3]
 [4 5 6]
 [7 8 9]]

array_two.ndim:2,array_two.shape:(3, 3)



array_three: 
[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]]

array_three.ndim:3,array_three.shape:(2, 2, 2)


**1) [,] 방식 -> view**

In [ ]:
print(f'array_two[1,1]: {array_two[1,1]}')
print(f'array_two[0,2]: {array_two[0,2]}')
print(f'array_two[-1,-3]: {array_two[-1,-3]}')

array_two[1,1]: 5
array_two[0,2]: 3
array_two[-1,-3]: 7


**2) [][] 방식 -> copy**

In [ ]:
print(f'array_two[1][1]: {array_two[1][1]}')
print(f'array_two[0][2]: {array_two[0][2]}')
print(f'array_two[-1][-3]: {array_two[-1][-3]}')

array_two[1][1]: 5
array_two[0][2]: 3
array_two[-1][-3]: 7


**Slicing 2D arrays**
- row-wise

In [ ]:
print(f'array_two: \n{array_two}\n')

print(f'array_two[0,:]: \n{array_two[0,:]}\n')
print(f'array_two[1,:]: \n{array_two[1,:]}\n')
print(f'array_two[2,:]: \n{array_two[2,:]}\n')

array_two: 
[[1 2 3]
 [4 5 6]
 [7 8 9]]

array_two[0,:]: 
[1 2 3]

array_two[1,:]: 
[4 5 6]

array_two[2,:]: 
[7 8 9]



**Slicing 2D arrays**
- column-wise

In [ ]:
print(f'Array two as reference: \n{array_two}\n')

print(f'array_two[:,0]: \n{array_two[:,0]}\n')
print(f'array_two[:,1]: \n{array_two[:,1]}\n')
print(f'array_two[:,2]: \n{array_two[:,2]}\n')

Array two as reference: 
[[1 2 3]
 [4 5 6]
 [7 8 9]]

array_two[:,0]: 
[1 4 7]

array_two[:,1]: 
[2 5 8]

array_two[:,2]: 
[3 6 9]



**Slicing 2D array**
- both axis

In [ ]:
print(f'array_two[0:2,0]: \n{array_two[0:2,0]}\n')
print(f'array_two[0:2,0:2]: \n{array_two[0:2,0:2]}\n')
print(f'array_two[1:3,1:3]: \n{array_two[1:3,1:3]}\n')

array_two[0:2,0]: 
[1 4]

array_two[0:2,0:2]: 
[[1 2]
 [4 5]]

array_two[1:3,1:3]: 
[[5 6]
 [8 9]]



**Slicing 2D arrays**
- with stride

In [ ]:
print(f'array_two[0,::1]: \n{array_two[0,::1]}\n')
print(f'array_two[0,::2]: \n{array_two[0,::2]}\n')

print(f'array_two[::1,0]: \n{array_two[::1,0]}\n')
print(f'array_two[::2,0]: \n{array_two[::2,0]}\n')

array_two[0,::1]: 
[1 2 3]

array_two[0,::2]: 
[1 3]

array_two[::1,0]: 
[1 4 7]

array_two[::2,0]: 
[1 7]



**Slicing 3D arrays**

In [ ]:
print(f'array_three: \n{array_three}\n')

print(f'array_three[0]:\n{array_three[0]}\n')
print(f'array_three[1]:\n{array_three[1]}\n')

array_three: 
[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]]

array_three[0]:
[[1 2]
 [3 4]]

array_three[1]:
[[5 6]
 [7 8]]



In [ ]:
print(f'array_three[0][0,:]:\n{array_three[0][0,:]}\n')
print(f'array_three[1][1,:]:\n{array_three[1][1,:]}\n')

array_three[0][0,:]:
[1 2]

array_three[1][1,:]:
[7 8]



## 3. Indexing and Assignment

In [ ]:
x = np.arange(10)
print(f'Array x:{x}\n')

x[2] = 33
print(f'x[2] = 33: {x}\n')

x[6:-1] = 728
print(f'x[6:-1] = 728: {x}')

Array x:[0 1 2 3 4 5 6 7 8 9]

x[2] = 33: [ 0  1 33  3  4  5  6  7  8  9]

x[6:-1] = 728: [  0   1  33   3   4   5 728 728 728   9]


In [ ]:
y = np.arange(10)
y[-6:-1] = np.arange(100,105)
print(f'y[-6:-1] = np.arange(100,105):\n\n{y}')

y[-6:-1] = np.arange(100,105):

[  0   1   2   3 100 101 102 103 104   9]


In [ ]:
z = np.arange(10)
z[2:5] = z[2:5]*27
print(f'z[2:5] = z[2:5]*27:\n\n{z}')

z[2:5] = z[2:5]*27:

[  0   1  54  81 108   5   6   7   8   9]


## 4. Boolean(Masking) Indexing

In [ ]:
array = np.arange(12)
mask = array > 6

print(f'array: \n{array}\n')
print(f'mask: \n{mask}\n')
print(f'array[mask]: \n{array[mask]}\n')

array: 
[ 0  1  2  3  4  5  6  7  8  9 10 11]

mask: 
[False False False False False False False  True  True  True  True  True]

array[mask]: 
[ 7  8  9 10 11]



In [ ]:
print(f'[(1 < array) & (array < 5)]:\n{[(1 < array) & (array < 5)]}\n')
print(f'array[(1 < array) & (array < 5)]:\n{array[(1 < array) & (array < 5)]}')

[(1 < array) & (array < 5)]:
[array([False, False,  True,  True,  True, False, False, False, False,
       False, False, False])]

array[(1 < array) & (array < 5)]:
[2 3 4]


In [ ]:
print(f'array[(2 == array) | (array > 9)]:\n{array[(2 == array) | (array > 9)]}\n')
print(f'array[((array % 2) == 0) | (array > 9)]:\n{array[((array % 2) == 0) | (array > 9)]}\n')
print(f'array[(2 != array) & (array != 7) & (array != 10)]:\n{array[(2 != array) & (array != 7) & (array != 10)]} ')

array[(2 == array) | (array > 9)]:
[ 2 10 11]

array[((array % 2) == 0) | (array > 9)]:
[ 0  2  4  6  8 10 11]

array[(2 != array) & (array != 7) & (array != 10)]:
[ 0  1  3  4  5  6  8  9 11] 


**2D array masking**
- mask 를 적용하고 난 후의 결과는?

In [ ]:
array_two = np.arange(1,17).reshape((4,4))
mask_two =  array_two > 8

print(f'array_two: \n{array_two}\n')
print(f'mask_two: \n{mask_two}\n')
print(f'array_two[mask_two]: \n{array_two[mask_two]}\n')
print(f'array_two[mask_two].shape: {array_two[mask_two].shape}')

array_two: 
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]]

mask_two: 
[[False False False False]
 [False False False False]
 [ True  True  True  True]
 [ True  True  True  True]]

array_two[mask_two]: 
[ 9 10 11 12 13 14 15 16]

array_two[mask_two].shape: (8,)


## 실습해 보기

In [ ]:
import numpy as np

#### 1. 다음 array 'a' 에 대해 원하는 결과를 얻도록 Indexing 해 보자.
- a = np.arange(10, 20)

1) 10  
2) 13  
3) 19  
4) array([10, 11])  
5) array([14, 15, 16])  
6) array([18, 19])  
7) array([10, 12, 14, 16, 18])  
8) array([11, 13, 15])

#### 2. 다음 array 'a' 에 대해 원하는 결과를 얻도록 Indexing 해 보자.
- a = np.arange(1,13).reshape(3,4)

1) 1  
2) 6  
3) 12  
4) array([1, 2, 3, 4])  
5) array([5, 6, 7, 8])  
6) array([ 9, 10, 11, 12])  
7) array([ 2,  6, 10])  
8) array([[1, 2],[5, 6]])  
9) array([5, 6])  
10) array([5, 7])

#### 3. 다음 array 'a' 를 변경해서 원하는 결과를 얻어보자.
- a = np.arange(10)

1) array([ 0,  1,  2,  3, 40, 50, 60,  7,  8,  9])

#### 4. 다음 array 'a' 를 변경해서 원하는 결과를 얻어보자.
- a = np.arange(1,10).reshape(3,3)

1) array([[  1,   2,   3],  
       [  4, 100, 101],  
       [  7, 103, 104]])

#### 4. 다음 array 'a' 가 아래 결과를 만족하도록 mask 를 만들고 boolean indexing 을 수행하라.
<br>
a = np.arange(1,11)

- [1,2,3,4,5]
- [2,4,6,,8,10]
- [1,2,3,4,6,8,10]